Pr 2

ltfscode opened the pull request.

GitEsense commented on 10-array-clean/app.js:
```diff
@@ -0,0 +1,15 @@
+  return num > 5;
+}
+
+console.log(reverseFilter([3, 6, 9, 2], deleteNumGreater));
```
тут всё "ок", вопросов нет)

GitEsense commented on 7-converter/app.js:
```diff
@@ -0,0 +1,73 @@
+        case "usd":
+          return sum;
+      }
+    
+    case "eur":
+      switch (outCurrency) {
+        case "rub":
+          return (sum * eur) / rub;
+
+        case "usd":
+          return (sum * eur) / usd;
+
+        case "cny":
+          return (sum * eur) / cny;
+
+        case "eur":
+          return sum;
+      }
+    
+    case "cny":
+      switch (outCurrency) {
+        case "rub":
+          return (sum * cny) / rub;
+
+        case "usd":
+          return (sum * cny) / usd;
+
+        case "eur":
+          return (sum * cny) / eur;
+
+        case "cny":
+          return sum;
+      }
+
+    default:
+      return null;
+  }
+}
+
+console.log(converter(2600000, 'eur', 'cny'));
```
Тут тоже всё отлично, но я бы немного "изменил", вот три разных варианта решения

**Простой SWITCH CASE:**
```js
function converter(sum, currency, targetCurrency) {
    const USD = 90;
    const EUR = 100;

    currency = currency.toUpperCase();
    targetCurrency = targetCurrency.toUpperCase();

    switch (currency) {
        case 'USD':
            switch (targetCurrency) {
                case 'RUB':
                    return sum * USD;
                case 'EUR':
                    return (sum * USD) / EUR;
                default:
                    return null;
            }
        case 'EUR':
            switch (targetCurrency) {
                case 'RUB':
                    return sum * EUR;
                case 'USD':
                    return (sum * EUR) / USD;
                default:
                    return null;
            }
        case 'RUB':
            switch (targetCurrency) {
                case 'USD':
                    return sum / USD;
                case 'EUR':
                    return sum / EUR;
                default:
                    return null;
            }
        default:
            return null;
    }
}

const convertResult = converter(1500, 'rub', 'usd');
console.log(convertResult);
```

**SWITCH CASE с проверками и шаблоном:**
```js
function convertSum(sum, from, to) {
    return ((sum * from) / to).toFixed(2);
}

function getTargetCurrencyIcon(targetCurrency) {
        switch (targetCurrency) {
            case 'RUB':
                return 'руб';
            case 'EUR':
                return '€';
            case 'USD':
                return '$';
            default: return '';
        }
}

function messageTemplate(value, icon){
    return `${value} ${icon}`;
}

function converter(sum, currency, targetCurrency) {
    const ALLOW_CURRENCY_LIST = ['RUB', 'EUR', 'USD'];
    const USD = 91.33;
    const RUB = 1;
    const EUR = 112.33;

    currency = currency.toUpperCase();
    targetCurrency = targetCurrency.toUpperCase();
    const icon = getTargetCurrencyIcon(targetCurrency);

    if (currency === targetCurrency) {
        return messageTemplate(sum, icon);
    }
    if (!ALLOW_CURRENCY_LIST.includes(currency)) {
        console.log('Входящая валюта мне неизвестна');
        return null;
    }
    if (!ALLOW_CURRENCY_LIST.includes(targetCurrency)) {
        console.log('Исходящая валюта мне неизвестна');
        return null;
    }
    
    let value = null;

    switch (currency) {
        case 'USD':
            switch (targetCurrency) {
                case 'RUB': value = convertSum(sum, USD, RUB); break;
                case 'EUR': value = convertSum(sum, USD, EUR); break;
                default: {
                    console.log('Исходящая валюта мне неизвестна');
                    return null;
                }
            }; break;
        case 'RUB':
            switch (targetCurrency) {
                case 'USD': value = convertSum(sum, RUB, USD); break;
                case 'EUR': value = convertSum(sum, RUB, EUR); break;
                default: {
                    console.log('Исходящая валюта мне неизвестна');
                    return null;
                }
            }; break;
        case 'EUR':
            switch (targetCurrency) {
                case 'RUB':
                    value = convertSum(sum, EUR, RUB); break;
                case 'USD':
                    value = convertSum(sum, EUR, USD); break;
                default: {
                    console.log('Исходящая валюта мне неизвестна');
                    return null;
                }
            }; break;
    }

    return messageTemplate(value, icon)
}
```

**Исключительно через массив:**
```js
function convertSum(sum, from, to) {
    return ((sum * from) / to).toFixed(2);
}

function getTargetCurrencyIcon(targetCurrency) {
        switch (targetCurrency) {
            case 'RUB':
                return 'руб';
            case 'EUR':
                return '€';
            case 'USD':
                return '$';
            default: return '';
        }
}

function messageTemplate(value, icon){
    return `${value} ${icon}`;
}

function converter(sum, currency, targetCurrency) {
    const ALLOW_CURRENCY_LIST = [['RUB', 1], ['EUR', 112.33], ['USD', 91.33]];
    const flatList = ALLOW_CURRENCY_LIST.flat();
    currency = currency.toUpperCase();
    targetCurrency = targetCurrency.toUpperCase();

    if (!flatList.includes(currency)) {
        console.log('Входящая валюта мне неизвестна');
        return null;
    }
    if (!flatList.includes(targetCurrency)) {
        console.log('Исходящая валюта мне неизвестна');
        return null;
    }
    const icon = getTargetCurrencyIcon(targetCurrency);
    if (currency === targetCurrency) {
        return messageTemplate(sum, icon);
    }

    const currentIndex = flatList.indexOf(currency) / 2;
    const targetIndex = flatList.indexOf(targetCurrency) / 2;
    const [, currentValue] = ALLOW_CURRENCY_LIST[currentIndex];
    const [, targetValue] = ALLOW_CURRENCY_LIST[targetIndex];
    const result = convertSum(sum, currentValue, targetValue)
    return messageTemplate(result, icon);
}
```

Ну и запросы для "вывода"
```js
console.log(converter(100, 'RUB', 'USD')); // 1.09 $
console.log(converter(100, 'RUB', 'EUR')); // 0.89 €
console.log(converter(1000, 'EUR', 'USD')); // 1229.94 $
console.log(converter(1000, 'USD', 'EUR')); // 813.05 €
console.log(converter(1000, 'EUR', 'RUB')); // 112330.00 руб

console.log(converter(1500, 'EUR', 'RUB'));
console.log(converter(1500, 'USD', 'RUB'));

console.log(converter(10, 'RUB', 'EUR'));
console.log(converter(10, 'USD', 'EUR'));

console.log(converter(10, 'RUB', 'EURU'));
console.log(converter(10, 'USD', 'EUR'));

console.log(converter(10, 'RUB', 'EUR'));
console.log(converter(10, 'USDTY', 'EUR'));
```

GitEsense commented on 8-crypto/app.js:
```diff
@@ -0,0 +1,38 @@
+}
+
+const originalPass = originalPassword(cryptoPass);
+console.log(originalPass === prompt("Введите ваш пароль"));
```
У нас должно быть две функции
1. Криптография (обратимая), т.е. если у нас пароль "password", то:
```js
const pwd = 'password';
const cryptPwd = cryptoPassword(pwd); // 'ssapdorw'
const encryptPwd = cryptoPassword(cryptPwd); // 'password'
```
2.  Функция "check", которая принимает в себя "оригинальный пароль" ('password') и шифрованный пароль ('ssapdorw') - должна вернуть результат сравнения оригинального пароля и расшифровки (через обратимость),
проще говоря выглядеть будет так:
```js
function checkPassword(encryptedPassword, originalPassword) {
    if (!encryptedPassword || !originalPassword) {
        return false;
    }
    return originalPassword === crypt(encryptedPassword);
}
```
И ничего более, так что функцию "шифрования" - надо переделать, чтобы была обратима

GitEsense commented on 9-sort-loops/app.js:
```diff
@@ -0,0 +1,28 @@
+arr = [1, 4, -5, 10, 0];
+function sortArrayLoopsInTwoLoops(arr, sortParam) {
+  let tempArr = [...arr];
+  let length = tempArr.length;
+
+  for (let i = 0; i < length - 1; i++) {
+    for (let j = 0; j < length - i - 1; j++) {
+      if (sortParam === "Возрастание") {
+        if (tempArr[j] > tempArr[j + 1]) {
+          let temp = tempArr[j];
+          tempArr[j] = tempArr[j + 1];
+          tempArr[j + 1] = temp;
+        }
+      } else if (sortParam === "Убывание") {
+        if (tempArr[j] < tempArr[j + 1]) {
+          let temp = tempArr[j];
+          tempArr[j] = tempArr[j + 1];
+          tempArr[j + 1] = temp;
+        }
+      }
+    }
+  }
+
+  return tempArr;
+}
+
+console.log(sortArrayLoopsInTwoLoops(arr, "Возрастание"));
+console.log(sortArrayLoopsInTwoLoops(arr, "Убывание"));
```
Тут всё отлично, но лучше несколько упростить и дать "значение" по умолчанию
```js
function orderTemplate(firstNumber, secondNumber, desc) {
    if (!desc) {
        return firstNumber > secondNumber;
    }
    return firstNumber < secondNumber;
}

function sortArrayInAscending(incomingArray, desc = false) {
    const array = [...incomingArray];
    for (let i = 0; i < array.length; i++) {
        for (let j = i + 1; j < array.length; j++) {
            const isExchange = orderTemplate(array[i], array[j], desc);
            if (isExchange) {
                [array[i], array[j]] = [array[j], array[i]];
            }
        }
    }
    return array;
}

console.log(numbers);
console.log('Отсортированный по возрастанию массив:', sortArrayInAscending(numbers));
console.log('Отсортированный по убыванию массив:', sortArrayInAscending(numbers, true));
```
или без отдельной функции
```js
function sortArrayInAscending(incomingArray, desc = false) {
    const array = [...incomingArray];
    for (let i = 0; i < array.length; i++) {
        for (let j = i + 1; j < array.length; j++) {
            const isExchange = desc ? array[i] < array[j] : array[i] > array[j];
            if (isExchange) {
                [array[i], array[j]] = [array[j], array[i]];
            }
        }
    }
    return array;
}

console.log(numbers);
console.log('Отсортированный по возрастанию массив:', sortArrayInAscending(numbers));
console.log('Отсортированный по убыванию массив:', sortArrayInAscending(numbers, true));
```

А вот функция генерации данных, чтобы можно было "поглядеть" на больших размерах массива
```js
const length = 1000; // кол-во чисел в массиве
const numbers = Array.from({ length }, () => {
    const max = 1000000;
    const min = 1000;
    const factor = Math.random() > 0.5 ? 1 : -1;
    const randomNumber = Math.ceil(Math.random() * (max - min)) + min;
    return randomNumber * factor;
});
```

GitEsense reviewed (changes requested):
задание 7 и 9 - всё ок, задание 8 надо переделать

GitEsense commented on 8-crypto/app.js:
```diff
@@ -0,0 +1,79 @@
+
+function cryptoPassword(password) {
+  return password.split('').reverse().join('');
+}
```
А Вы бы сами доверили такой функции шифрования пароля свой пароль ?)
С точки зрения "ТЗ" конечно верно, но с точки зрения реализации - халтура))
 Вот как простой пример для анализа
```js
/**
 * function encrypt / decrypt string password
 * @param {string} password - A string password
 */
function crypt(password) {
    if (!password || !password.length) {
        return NaN;
    }
    const splitedArray = password.split('');
    if (splitedArray.length < 8) {
        return false;
    }

    const middleIndex = Math.floor(splitedArray.length / 2);
    const firstHalf = splitedArray.slice(0, middleIndex).reverse();
    const secondHalf = splitedArray.slice(middleIndex).reverse();
    const length = secondHalf.length;
    [secondHalf[1], secondHalf[length - 2]] = [secondHalf[length - 2], secondHalf[1]];

    const encryptedPassword = firstHalf.concat(secondHalf);
    return encryptedPassword.join('');
}

function checkPassword(encryptedPassword, originalPassword) {
    if (!encryptedPassword || !originalPassword) {
        return false;
    }
    return originalPassword === crypt(encryptedPassword);
}

function getRandomNumber(min = 0, max = 86) {
    return Math.ceil(Math.random() * (max - min)) + min;
}

function getRandomPassword(length = 12) {
    const symbolArray = Array.from({ length: 87 }, (_, index) => String.fromCharCode(index + 33));
    console.log(symbolArray);
    const password = Array.from({ length }, () => symbolArray[getRandomNumber()]);
    return password.join('');
}

/*
    logging and test function with console
*/

const password = getRandomPassword();
const encrypted = crypt(password);

console.log('Пароль:', password);
console.log('Зашифрованный пароль:', encrypted);
console.log('Проверка пароля:', checkPassword(encrypted, password));

const password2 = getRandomPassword();
const encrypted2 = crypt(password2).substring(1);

console.log('Пароль:', password2);
console.log('Зашифрованный пароль:', encrypted2);
console.log('Проверка пароля:', checkPassword(encrypted2, getRandomPassword()));
```

ltfscode commented:
Спасибо. Пока ещё не всегда могу понять где и что применять из пройденного

вт, 4 июн. 2024 г., 22:37 kld_ch_a ***@***.***>:

> ***@***.**** approved this pull request.
> ------------------------------
>
> In 8-crypto/app.js
> <https://github.com/ltfscode/dz-js/pull/2#discussion_r1626516545>:
>
> > +//   return passwordArr;
> +// }
> +
> +// function checkPassword(encryptedPassword, originalPassword) {
> +//   if (!encryptedPassword || !originalPassword) {
> +//     return false;
> +//   }
> +//   return originalPassword === decryptPassword(encryptedPassword);
> +// }
> +
> +// console.log(checkPassword(prompt("Введите ваш пароль"), prompt("Введите ваш пароль")));
> +
> +
> +function cryptoPassword(password) {
> +  return password.split('').reverse().join('');
> +}
>
> А Вы бы сами доверили такой функции шифрования пароля свой пароль ?)
> С точки зрения "ТЗ" конечно верно, но с точки зрения реализации - халтура))
> Вот как простой пример для анализа
>
> /** * function encrypt / decrypt string password * @param {string} password - A string password */function crypt(password) {
>     if (!password || !password.length) {
>         return NaN;
>     }
>     const splitedArray = password.split('');
>     if (splitedArray.length < 8) {
>         return false;
>     }
>
>     const middleIndex = Math.floor(splitedArray.length / 2);
>     const firstHalf = splitedArray.slice(0, middleIndex).reverse();
>     const secondHalf = splitedArray.slice(middleIndex).reverse();
>     const length = secondHalf.length;
>     [secondHalf[1], secondHalf[length - 2]] = [secondHalf[length - 2], secondHalf[1]];
>
>     const encryptedPassword = firstHalf.concat(secondHalf);
>     return encryptedPassword.join('');}
> function checkPassword(encryptedPassword, originalPassword) {
>     if (!encryptedPassword || !originalPassword) {
>         return false;
>     }
>     return originalPassword === crypt(encryptedPassword);}
> function getRandomNumber(min = 0, max = 86) {
>     return Math.ceil(Math.random() * (max - min)) + min;}
> function getRandomPassword(length = 12) {
>     const symbolArray = Array.from({ length: 87 }, (_, index) => String.fromCharCode(index + 33));
>     console.log(symbolArray);
>     const password = Array.from({ length }, () => symbolArray[getRandomNumber()]);
>     return password.join('');}
> /*    logging and test function with console*/
> const password = getRandomPassword();const encrypted = crypt(password);
> console.log('Пароль:', password);console.log('Зашифрованный пароль:', encrypted);console.log('Проверка пароля:', checkPassword(encrypted, password));
> const password2 = getRandomPassword();const encrypted2 = crypt(password2).substring(1);
> console.log('Пароль:', password2);console.log('Зашифрованный пароль:', encrypted2);console.log('Проверка пароля:', checkPassword(encrypted2, getRandomPassword()));
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ltfscode/dz-js/pull/2#pullrequestreview-2097243951>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BETJP2JJ6J57ICVCRVN6SGDZFYJPZAVCNFSM6AAAAABIRNT27KVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAOJXGI2DGOJVGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
